ENH: support bytearray and unicode (#3)

westurner opened the pull request:
- #3 
- [x] TST: manual test bencode.dumps with Python 2
- [x] TST: manual test bencode.dumps with Python 3
- [x] TST: test that a .torrent file at least loads (with Transmission)
  - [ ] is this handling of utf8 unicode strings consistent with which BEP?
    - https://wiki.theory.org/BitTorrentSpecification#Metainfo_File_Structure
      " All character string values are UTF-8 encoded."
- [x] TST: write test cases for bytearrays
- [ ] TST: write test cases for unicodes
- [ ] PRF: port to _fast.c

westurner committed:
ENH: _pure.py: support bytearray and unicode (#3)

notpushkin commented:
Fails on Python 3, `NameError: name 'unicode' is not defined` :(

westurner commented:
Whoops. I think that should only be for Python 2 because in Python 3 all
str are unicode.

On Wednesday, November 16, 2016, Alexander Pushkov notifications@github.com
wrote:

> Fails on Python 3, NameError: name 'unicode' is not defined :(
> 
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> https://github.com/kosqx/better-bencode/pull/4#issuecomment-260935205,
> or mute the thread
> https://github.com/notifications/unsubscribe-auth/AADGy77ug9XCD5PdA461pQNqM3AtSJzmks5q-vfTgaJpZM4Kwxlh
> .

notpushkin commented:
The error actually happens in [the `elif t is unicode`](https://github.com/westurner/better-bencode/blob/62d4d89c79588629546fd24df8d557d85f097d28/better_bencode/_pure.py#L66) line. We can safely substitute unicode for str (in py3k) here (as `unicode_to_bytes` accepts str in python 3), so I suggest this simple fix: https://github.com/Songbee/better-bencode/commit/0bb114b9875fac1a6829e0a7bd8ffb933c66c525

westurner committed:
BUG: _pure.py: unicode = str for py3k

As suggested by
https://github.com/Songbee/better-bencode/commit/0bb114b9875fac1a6829e0a7bd8ffb933c66c525

westurner commented:
Good call @iamale

westurner committed:
TST: test_bencode.py: remove u'' from TESTS_TYPEERROR

westurner committed:
TST: test_bencode.py: add bytearrays to TEST_DATA

note: this passes w/ _pure.py but not w/ _fast.c

westurner commented:
- [ ] TST: What should I add to `TEST_DATA` to test for unicode strings?
- [ ] ENH,TST: IDK Python C-API well enough to port this from _pure.py to _fast.c

westurner commented:
...  TIL "All character string values are UTF-8 encoded." https://wiki.theory.org/BitTorrentSpecification#Metainfo_File_Structure

westurner commented on tests/test_bencode.py:
```diff
@@ -65,6 +65,8 @@ def test_error_dump(module):
     (b'd3:bar4:spam3:fooi42ee', {b'bar': b'spam', b'foo': 42}),
     (b'd1:ai1e1:bi2e1:ci3ee', {b'a': 1, b'b': 2, b'c': 3}),
     (b'd1:a1:be', {b'a': b'b'}),
+    (b'3:\x00\x01\x02', bytearray([0, 1, 2])),
```
is this correct?

notpushkin commented:
Also I think it would be nice to decode unicode strings to `unicode`/`str` as well (although it might be a bit trickier).
